aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -576,14 +681,26 @@ namespace {
   }
 
   std::vector<int64_t> _hidden_size(const RNNDescriptorParams& rnn, const TensorDescriptorListParams& tensors) {
+    if (rnn.proj_size != 0 && rnn.proj_size != rnn.hidden_size) {
```

Comment:
you are checking rnn.proj_size != rnn.hidden_size in multiple places, but in rnn.py you explicitly assert that proj_size is less than hidden_size, so it can never be true?

Reply:
So, there are actually 2 ways how "no projection" LSTM can currently be defined. First, if proj_size is explicitly set to 0. This way assumes that proj_size is passed directly as an argument to a function and not inferred from cell dimensions. But in fact the main point of entry to all these functions is through "_cudnn_impl" function, which infers "proj_size" from hidden dimensions. In that case was easier for me to just say that "proj_size = hx.size(2); hidden_size = cs.size(2)" and if they are not equal, we will have LSTM with projections, if they are equal, it's usual LSTM. I agree that it may be misleading (which you also point out in the later comment). Let me try to unify this convention and always explicitly assign proj_size to 0 when it is inferred and hidden dimensions are equal